core/runtime_modules.rs

```diff
@@ -0,0 +1,206 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::error::AnyError;
+use crate::{OpFn, OpId, OpState};
+
+pub type SourcePair = (&'static str, &'static str);
+pub type OpPair = (&'static str, Box<OpFn>);
+pub type RcOpRegistrar = Rc<RefCell<dyn OpRegistrar>>;
+pub type OpStateFn = dyn Fn(&mut OpState) -> Result<(), AnyError>;
+
+// JsRuntimeModule defines a common interface followed by all op_crates
+// so the JsRuntime can handle initialization consistently (e.g: snapshots or not)
+// this makes op_crates plug-n-play, since modules can be simply passed to a JsRuntime:
+// ```
+// JsRuntime:new(RuntimeOptions{
+//  modules: vec![deno_url::init(), deno_webgpu::init(), ... ],
+//  ..Default::default()
+// })
+pub trait JsRuntimeModule {
+  /// This function returns JS source code to be loaded into the isolate (either at snapshotting, or at startup).
+  /// as a vector of a tuple of the file name, and the source code.
+  fn init_js(&self) -> Result<Vec<SourcePair>, AnyError> {
+    // default implementation of `init_js` is to load no runtime code
+    Ok(vec![])
+  }
+
+  /// This function can set up the initial op-state of an isolate at startup.
+  fn init_state(&self, _state: &mut OpState) -> Result<(), AnyError> {
+    // default implementation of `init_state is to not mutate the state
+    Ok(())
+  }
+
+  /// This function lets you middleware the op registrations. This function gets called before this module's init_ops.
+  fn init_op_registrar_middleware(
+    &self,
+    registrar: RcOpRegistrar,
+  ) -> RcOpRegistrar {
+    // default implementation is to not change the registrar
+    registrar
+  }
+
+  /// This function gets called at startup to initialize the ops in the isolate.
+  fn init_ops(&mut self, _registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    // default implementation of `init_ops` is to load no runtime code
+    Ok(())
+  }
+}
+
+// A simple JsRuntimeModule
+pub struct BasicModule {
+  js_files: Option<Vec<SourcePair>>,
+  ops: Option<Vec<OpPair>>,
+  opstate_fn: Option<Box<OpStateFn>>,
+}
+
+impl BasicModule {
+  pub fn new(
+    js_files: Option<Vec<SourcePair>>,
+    ops: Option<Vec<OpPair>>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
+    Self {
+      js_files,
+      ops,
+      opstate_fn,
+    }
+  }
+
+  pub fn pure_js(js_files: Vec<SourcePair>) -> Self {
+    Self::new(Some(js_files), None, None)
+  }
+
+  pub fn with_ops(
+    js_files: Vec<SourcePair>,
+    ops: Vec<OpPair>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
```

Comment:
Seems superfluous. Does this offer much more than `new()`? Using a couple fewer `Some` wrappers is offset by the mental overhead of having to understand all these different constructors.

Reply:
@ry Not really, especially given that `vec![]` isn't terribly more verbose than `None` so `js_files` and `ops` could just be `Vec`s and not `Option`s in the struct.

I might introduce a builder pattern, something like:
```rust
BasicExtension::builder()
   .js(...)
   .ops(...)
   .state(move || { ... })
   .ops(...) // More ops to group by wrapper and "avoid" `declare_ops_group` ...
   .build()
```